// @corpsmap/corpsmap/src/application-bundles/plugin-bundle.js
import { createSelector } from "redux-bundler";
import _ from "lodash";

export default {
  name: "plugins",

  getReducer() {
    const initialData = {
      loading: false,
      components: [],
    };

    return (state = initialData, { type, payload }) => {
      if (type === "PLUGINS_STARTED_LOADING") {
        return Object.assign({}, state, {
          loading: true,
        });
      }

      if (type === "PLUGINS_STOPPED_LOADING") {
        return Object.assign({}, state, {
          loading: false,
        });
      }

      if (type === "PLUGINS_ADDED_COMPONENTS") {
        return Object.assign({}, state, {
          components: state.components.concat(payload),
        });
      }

      return state;
    };
  },

  // ACTION CREATORS
  doAddPluginComponents: (components) => ({ dispatch }) => {
    dispatch({ type: "PLUGINS_ADDED_COMPONENTS", payload: components });
  },

  doStartLoadingPlugins: () => ({ dispatch, store }) => {
    dispatch({ type: "PLUGINS_STARTED_LOADING" });
  },

  doStopLoadingPlugins: () => ({ dispatch }) => {
    dispatch({ type: "PLUGINS_STOPPED_LOADING" });
  },

  // SELECTORS
  selectPluginsAreLoading: (state) => {
    return state.plugins.loading;
  },

  selectAllComponents: (state) => {
    return state.plugins.components;
  },

  selectComponentsLeftSidebarTop: createSelector(
    "selectAllComponents",
    (allComponents) => {
      return _.map(
        _.sortBy(_.filter(allComponents, { region: "left-sidebar-top" }), [
          "sortOrder",
        ]),
        (c) => {
          return c.component;
        }
      );
    }
  ),

  selectComponentsLeftSidebarBottom: createSelector(
    "selectAllComponents",
    (allComponents) => {
      return _.map(
        _.sortBy(_.filter(allComponents, { region: "left-sidebar-bottom" }), [
          "sortOrder",
        ]),
        (c) => {
          return c.component;
        }
      );
    }
  ),

  selectComponentsLeftSidebarCount: createSelector(
    "selectComponentsLeftSidebarTop",
    "selectComponentsLeftSidebarBottom",
    (top, bottom) => {
      return top.length + bottom.length;
    }
  ),

  selectComponentsRightSidebarTop: createSelector(
    "selectAllComponents",
    (allComponents) => {
      return _.map(
        _.sortBy(_.filter(allComponents, { region: "right-sidebar-top" }), [
          "sortOrder",
        ]),
        (c) => {
          return c.component;
        }
      );
    }
  ),

  selectComponentsRightSidebarBottom: createSelector(
    "selectAllComponents",
    (allComponents) => {
      return _.map(
        _.sortBy(_.filter(allComponents, { region: "right-sidebar-bottom" }), [
          "sortOrder",
        ]),
        (c) => {
          return c.component;
        }
      );
    }
  ),

  selectComponentsRightSidebarCount: createSelector(
    "selectComponentsUpperRight",
    "selectComponentsLowerRight",
    "selectComponentsRightSidebarTop",
    "selectComponentsRightSidebarBottom",
    (ur, lr, top, bottom) => {
      return ur.length + lr.length + top.length + bottom.length;
    }
  ),

  selectComponentsBasicToolbar: createSelector(
    "selectAllComponents",
    (allComponents) => {
      return _.map(
        _.sortBy(_.filter(allComponents, { region: "basic-toolbar" }), [
          "sortOrder",
        ]),
        (c) => {
          return c.component;
        }
      );
    }
  ),

  selectComponentsBasicToolbarCount: createSelector(
    "selectComponentsBasicToolbar",
    (components) => {
      return components.length;
    }
  ),

  selectComponentsLowerRight: createSelector(
    "selectAllComponents",
    (allComponents) => {
      return _.map(
        _.sortBy(_.filter(allComponents, { region: "lower-right-toolbar" }), [
          "sortOrder",
        ]),
        (c) => {
          return c.component;
        }
      );
    }
  ),

  selectComponentsUpperRight: createSelector(
    "selectAllComponents",
    (allComponents) => {
      return _.map(
        _.sortBy(_.filter(allComponents, { region: "upper-right-toolbar" }), [
          "sortOrder",
        ]),
        (c) => {
          return c.component;
        }
      );
    }
  ),
};
