// @corpsmap/corpsmap/src/core-plugins/measure-tools/measure-options.js
import React from "react";
import { connect } from "redux-bundler-react";

class MeasureToolOptions extends React.Component {
  constructor(props) {
    super(props);
    this.updateUnits = this.updateUnits.bind(this);
    this.close = this.close.bind(this);
  }

  close() {
    const { doModalClose } = this.props;
    doModalClose();
  }

  updateUnits = e => {
    const {
      doUpdateMeasureUnits,
      measureAreaUnits,
      measureDistanceUnits
    } = this.props;
    const val = e.currentTarget.value;
    const type = e.currentTarget.dataset.type;
    const units = {
      areaUnits: measureAreaUnits,
      distanceUnits: measureDistanceUnits
    };
    units[type] = val;
    doUpdateMeasureUnits(units);
  };

  render() {
    const { theme, measureAreaUnits, measureDistanceUnits } = this.props;
    const height = 300;
    return (
      <div>
        <div className={"cm-modal-header " + theme}>Measure Tool Options</div>
        <div
          className="cm-modal-body"
          style={{ height: height, top: 0, bottom: 0, left: 0, right: 0 }}
        >
          <div className="cm-form-group">
            <label className="cm-control-label">Distance Units</label>
            <select
              onChange={this.updateUnits}
              value={measureDistanceUnits}
              data-type="distanceUnits"
              className="cm-form-control"
              id="select"
            >
              <option value="ft/mi">Feet / Miles</option>
              <option value="m/km">Meters / Kilometers</option>
            </select>
          </div>

          <div className="cm-form-group">
            <label className="cm-control-label">Area Units</label>
            <select
              onChange={this.updateUnits}
              value={measureAreaUnits}
              data-type="areaUnits"
              className="cm-form-control"
              id="select"
            >
              <option value="ft/mi">Sq. Feet / Sq. Miles</option>
              <option value="m/km">Sq. Meters / Sq. Kilometers</option>
              <option value="acres">Acres</option>
              <option value="hectares">Hectares</option>
            </select>
          </div>
        </div>
        <div className="cm-modal-footer">
          <button
            className={"cm-btn cm-btn-theme " + theme}
            onClick={this.close}
          >
            Close
          </button>
        </div>
      </div>
    );
  }
}

export default connect(
  "doUpdateMeasureUnits",
  "doModalClose",
  "selectTheme",
  "selectMeasureAreaUnits",
  "selectMeasureDistanceUnits",
  MeasureToolOptions
);
