// @corpsmap/corpsmap/src/core-plugins/draw-tools/draw-styles.js

import Style from 'ol/style/Style';
import Circle from 'ol/style/Circle';
import Fill from 'ol/style/Fill';
import Stroke from 'ol/style/Stroke';
import Text from 'ol/style/Text';

export default (type) => {

  const fillColor = 'rgba(245, 66, 233, 1)';
  const fillColorA = 'rgba(245, 66, 233, 0.6)';
  const strokeColor = '#ffffff';

  const styles = {
    'Point': (f) => {
      const label = f.get('label') || '';
      return new Style({
        image: new Circle({
          radius: 7,
          fill: new Fill({
            color: fillColor
          }),
          stroke: new Stroke({
            color: strokeColor,
            width: 3
          }),
        }),
        text: new Text({
          placement: 'point',
          font: '14px sans-serif',
          weight: 'bold',
          text: label,
          textAlign: 'left',
          textBaseline: 'bottom',
          offsetX: 4,
          offsetY: -4,
          overflow: true,
          stroke: new Stroke({
            color: strokeColor,
            width: 3
          })
        }),
        fill: new Fill({
          color: fillColor
        }),
        stroke: new Stroke({
          color: strokeColor,
          width: 3
        })
      })
    },
    'LineString': (f) => {
      const label = f.get('label') || '';
      return [
        new Style({
          stroke: new Stroke({
            color: strokeColor,
            width: 4
          })
        }),
        new Style({
          text: new Text({
            font: '14px sans-serif',
            weight: 'bold',
            text: label,
            textAlign: 'center',
            textBaseline: 'middle',
            placement: 'line',
            overflow: true,
            stroke: new Stroke({
              color: strokeColor,
              width: 3
            })
          }),
          stroke: new Stroke({
            color: fillColor,
            width: 2
          })
        })
      ]
    },
    'Polygon': (f) => {
      const label = f.get('label') || '';
      return new Style({
        text: new Text({
          placement: 'point',
          font: '14px sans-serif',
          weight: 'bold',
          text: label,
          textAlign: 'center',
          textBaseline: 'middle',
          overflow: true,
          stroke: new Stroke({
            color: strokeColor,
            width: 3
          })
        }),
        fill: new Fill({
          color: fillColorA
        }),
        stroke: new Stroke({
          color: strokeColor,
          width: 3
        })
      })
    }
  }

  return styles[type];
}
