// @corpsmap/corpsmap/src/core-plugins/editor/date-editor.js
import DatePicker from 'react-datepicker';
import moment from 'moment';

import React from 'react';
import EditorBase from './editor-base';

import 'react-datepicker/dist/react-datepicker.css';

class DateEditor extends EditorBase {

  constructor(props){
    super(props);

    const date = props.value ? moment(props.value) : moment();

    this.state = {
      key: props.column.key,
      date: date,
      val: date.format('YYYY-MM-DDTHH:mm:ss.SSSZZ')
    }

    this.handleChange = this.handleChange.bind(this);
    this.handleBlur = this.handleBlur.bind(this);
    this.getValue = this.getValue.bind(this);
  }

  handleChange(date){
    this.setState({
      date: date,
      val: date.format('YYYY-MM-DDTHH:mm:ss.SSSZZ')
    })
  }

  handleBlur(e){
    const { onBlur } = this.props;
    const date = moment(e.currentTarget.value);
    e.currentTarget.value = date.format('YYYY-MM-DDTHH:mm:ss.SSSZZ')
    onBlur(e);
  }

  getValue(){
    const { key, val } = this.state;
    return {[key]:val};
  }

  render(){
    return (
      <DatePicker
        ref={(el) => { this.input = el }}
        selected={this.state.date}
        onChange={this.handleChange}
        onBlur={this.props.onBlur}
        showTimeSelect={true}
        timeFormat="HH:mm"
        timeIntervals={30}
        dateFormat="LLL"
        timeCaption="time"
      />
    )
  }
}

export default DateEditor;
