// @corpsmap/corpsmap/src/core-plugins/editor/number-editor.js
import React from 'react';
import EditorBase from './editor-base';

class NumberEditor extends EditorBase {

  constructor(props){
    super(props);

    this.state = {
      val: props.value,
      size: props.column.size,
      scale: props.column.scale
    }

    this.handleChange = this.handleChange.bind(this);
    this.handleBlur = this.handleBlur.bind(this);
  }

  handleChange(e){
    const { size, scale } = this.state;
    let val = e.currentTarget.value;

    if(!Number(val)) return; //<-- if you give me NaN then I bail

    const hasDecimal = val.indexOf('.') !== -1;
    let valSize = hasDecimal ? val.length - 1 : val.length;
    if(valSize > size) return; //<-- if you give me too much then I bail

    if(scale === '0' && hasDecimal) return; //<-- if you try to give me decimals and i dont want them then I bail

    this.setState({
      val: val
    })
  }

  handleBlur(e){
    const { onBlur } = this.props;
    e.currentTarget.value = Number(e.currentTarget.value);
    onBlur(e);
  }


  render(){
    const { val } = this.state;
    return (
      <input 
        type="number"
        ref={(el) => { this.input = el }}
        value={ val }
        onBlur={this.handleBlur}
        onChange={this.handleChange}
      />
    )
  }
}

export default NumberEditor;
