// @corpsmap/corpsmap/src/core-plugins/zoom-home/zoom-home-button.js
import React from "react";
import { connect } from "redux-bundler-react";
import { SidebarButton } from "../../application-controls/index";

class ZoomHomeButton extends React.Component {
  constructor(props) {
    super(props);
    this.clock = null;
    this.handleClick = this.handleClick.bind(this);
    this.updateLoc = this.updateLoc.bind(this);
    this.startClock = this.startClock.bind(this);
    this.stopClock = this.stopClock.bind(this);
  }

  updateLoc() {
    const { doZoomHomeResetLoc, doNotificationsFire } = this.props;
    doZoomHomeResetLoc();
    doNotificationsFire({
      title: "Home Location Set",
      message: "Home button reset to current location",
      level: "success"
    });
  }

  handleClick() {
    const { doZoomHome } = this.props;
    doZoomHome();
  }

  startClock() {
    this.clock = window.setTimeout(this.updateLoc, 1000);
  }

  stopClock() {
    if (this.clock) window.clearTimeout(this.clock);
  }

  render() {
    return (
      <SidebarButton
        iconClass="mdi mdi-home"
        label="Zoom to initial extent"
        title="Zoom to initial extent"
        onClick={this.handleClick}
        onMouseDown={this.startClock}
        onMouseUp={this.stopClock}
      />
    );
  }
}

export default connect(
  "doZoomHome",
  "doZoomHomeResetLoc",
  "doNotificationsFire",
  ZoomHomeButton
);
