// @corpsmap/corpsmap/src/core-plugins/tree-view/layer-subpanel/opacity-slider.js
import React from "react";

class OpacitySlider extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      opacity: props.layer.getOpacity()
    };
    this.handleChange = this.handleChange.bind(this);
  }

  handleChange(e) {
    const { layer } = this.props;
    const val = e.currentTarget.value;
    this.setState(
      {
        opacity: Number(val)
      },
      () => {
        if (isNaN(this.state.opacity)) return;
        layer.setOpacity(this.state.opacity);
      }
    );
  }

  render() {
    const { opacity } = this.state;
    const formattedOpacity = "" + Math.round(opacity * 100) + "%";
    return (
      <div>
        <hr />
        <div className="cm-form-group">
          <label className="cm-control-label">
            Opacity: {formattedOpacity}
          </label>
          <input
            className="cm-input"
            value={opacity}
            onChange={this.handleChange}
            type="range"
            min={0}
            max={1}
            step={0.01}
          ></input>
        </div>
      </div>
    );
  }
}

export default OpacitySlider;
