// @corpsmap/corpsmap/src/application-controls/search-input.js
import React from "react";

class SearchInput extends React.Component {
  constructor(props) {
    super(props);
    this.updateFilterString = this.updateFilterString.bind(this);
    this.handleKeyUp = this.handleKeyUp.bind(this);
    this.setFocus = this.setFocus.bind(this);
  }
  updateFilterString(e) {
    const { onUpdate } = this.props;
    const val = e.currentTarget.value;
    onUpdate(val);
  }
  handleKeyUp(e) {
    const { onKeyUp } = this.props;
    if (onKeyUp) onKeyUp(e);
  }
  setFocus() {
    this.input.focus();
  }
  render() {
    const { filterString, placeholder, size, inline, darkTheme } = this.props;
    let inputClass = "cm-form-control";
    inputClass += size === "small" ? " input-sm" : "";
    const iconSize = size === "small" ? "17px" : "22px";
    const inputIndent = size === "small" ? "20px" : "30px";
    return (
      <div className="cm-form-group" style={inline ? { marginBottom: 0 } : {}}>
        <span style={{ position: "absolute", fontSize: iconSize, top: "10px" }}>
          <i className="mdi mdi-magnify" />
        </span>
        <span>
          <input
            ref={(el) => {
              this.input = el;
            }}
            style={{
              textIndent: inputIndent,
              color: darkTheme ? "#ffffff" : "inherit",
            }}
            value={filterString}
            type="text"
            className={inputClass}
            placeholder={placeholder}
            onChange={this.updateFilterString}
            onKeyUp={this.handleKeyUp}
          />
        </span>
      </div>
    );
  }
}

export default SearchInput;
