// @corpsmap/corpsmap/src/application-controls/sidebar-panel-opener.js
import React from "react";
import { connect } from "redux-bundler-react";
import SidebarButton from "./sidebar-button";

class SidebarPanelOpenerButton extends React.Component {
  constructor(props) {
    super(props);
    this.handleClick = this.handleClick.bind(this);
  }

  handleClick() {
    const {
      doPrimaryPanelOpen,
      doPrimaryPanelClose,
      title,
      panelComponent,
      panelOptions,
      primaryPanelOpenedBy
    } = this.props;
    if (primaryPanelOpenedBy !== title) {
      doPrimaryPanelOpen({
        openedBy: title,
        options: panelOptions,
        children: [panelComponent]
      });
    } else {
      doPrimaryPanelClose();
    }
  }

  render() {
    const {
      theme,
      label,
      title,
      iconClass,
      iconText,
      styleOverride,
      primaryPanelOpenedBy
    } = this.props;
    return (
      <SidebarButton
        active={primaryPanelOpenedBy === title}
        iconClass={iconClass}
        iconText={iconText}
        styleOverride={styleOverride}
        expanded={false}
        label={label}
        title={title}
        onClick={this.handleClick}
        theme={theme}
      />
    );
  }
}

export default connect(
  "selectTheme",
  "selectPrimaryPanelOpenedBy",
  "doPrimaryPanelOpen",
  "doPrimaryPanelClose",
  SidebarPanelOpenerButton
);
